rfcs/input-validations.md

```diff
@@ -2,160 +2,206 @@
     "role": "<role-name>",
     "session_variables": {
         "x-hasura-user-id": "<session-user-id>",
         "x-hasura-user-name": "<session-user-name>"
     },
     "data": {
-        "objects": [
-            {"column_1": "column_1_value", "column_2": "column_2_value", "relationship": [{"relationship_column_1": "column_value"}]},
-            {"column_1": "column_1_value", "column_2": "column_2_value", "relationship": [{"relationship_column_1": "column_value"}]}
-        ]
+        "objects": [JSON-fied <model_name>_insert_input!]
     }
 }
 ```
-- `version`: An integer version serves to indicate the request format. Whenever a breaking update occurs in the request
-payload, the version will be incremented. The initial version is set to `1`.
+
+- `version`: An integer version serves to indicate the request format. Whenever a
+breaking update occurs in the request payload, the version will be incremented. The
+initial version is set to `1`.
 - `role`: Hasura session role on which permissions are enforced.
-- `session_variables`: Session variables that aid in enforcing permissions. Variable names always starts with `x-hasura-*`.
-- `data.objects`: List of rows to be inserted which are specified in the `objects` input field of insert mutation. Also includes nested data of relationships.
+- `session_variables`: Session variables that aid in enforcing permissions. Variable
+  names always starts with `x-hasura-*`.
+- `data.objects`: List of rows to be inserted which are specified in the `objects`
+input field of insert mutation. Also includes nested data of relationships. The
+structure of this field will be similar to the JSONified structure of the
+`<model_name>_insert_input!` graphql type.
+
+Note that, in `data.objects` if the data to be inserted contains nested inserts, then the `data.objects` for the:
+1. The Root model has a type of `JSON-fied <model_name>_insert_input!`, i.e the nested inserts will be present as 
+`JSON-fied <model_name>_(arr|obj)_rel_insert_input!`
+2. The nested inserts payload has the type `JSON-fied <nested_model_name>_insert_input!`
 
 ### Response
 
-The HTTP handler should always return the JSON object with `200` status code. The object should contain `is_valid` field
-whose value is a boolean and an optional `error` field with a message forwarded to client when `is_valid` is `false`.
+1. Successfull Response
+
+The HTTP validation URL should return a `200` status code to represent successful
+validation.
 
 ```http
 200 OK
-
-{
-    "is_valid": true
-}
 ```
-or
+
+2. Unsucessfull Response
+
+The HTTP validation URL should return a `400` status code to represent failed
+validation. If any optional JSON object is returned then it will be forwarded to the
+client and will be included inside the `details` section of response.
+
+
 ```http
-200 OK
+400 BAD REQUEST
 
 {
-    "is_valid": false,
-    "error": "Phone number is invalid"
+    "message": "Validation Failed",
+    "details": {
```

Comment:
Where will `details` go in the graphql error?

Reply:
Thanks! As disucssed, we now only expect `message` field in the optional JSON object. If the user does not send any JSON object then there would be no message field in the final graphql error.

This is still okay, because the `extension.code` will have the value `input-validation-failed`.